## Content

## 01. Import libraries
## 02. Import data
## 03. Exercise - Part 2

## 01. Import libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path

path = r'C:\Users\Frederick\Documents\07-12-2023 Instacart Basket Analysis'

## 02. Import data

In [3]:
# Import data for orders_products merged with new variables added

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_added_variables.pkl'))

## 03. Exercise - Part 2

In [4]:
# Checking output of the ords_prods_merge dataframe

ords_prods_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  user_order_number  \
0             0             0   2539329        1                  1   
1             1             1   2398795        1                  2   
2             2             2    473747        1                  3   
3             3             3   2254736        1                  4   
4             4             4    431534        1                  5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   3                  7                    15.0         196   
2                   3                 12                    21.0         196   
3                   4                  7                    29.0         196   
4                   4                 15                    28.0         196   

   add_to_cart_order  ...  Unnamed: 0_y product_name  aisle_id department_id  \
0                  1  ...           195         Soda        77             7   
1                  1  ...           195         Soda        77             7   
2                  1  ...           195         Soda        77             7   
3                  1  ...           195         Soda        77             7   
4                  1  ...           195         Soda        77             7   

   prices  exists     busiest_day    price_range_loc  busiest_days  \
0     9.0    both  Regularly busy  Mid-range product  Regular days   
1     9.0    both  Regularly busy  Mid-range product  Slowest days   
2     9.0    both  Regularly busy  Mid-range product  Slowest days   
3     9.0    both      Least busy  Mid-range product  Slowest days   
4     9.0    both      Least busy  Mid-range product  Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders  

[5 rows x 22 columns]

In [5]:
# Q2. Find aggregated mean of 'user_order_number' column grouped by department_id 

ords_prods_merge.groupby('department_id').agg({'user_order_number': ['mean']})

user_order_number
                           mean
department_id                  
1                     15.457838
2                     17.277920
3                     17.170395
4                     17.811403
5                     15.215751
6                     16.439806
7                     17.225802
8                     15.340650
9                     15.895474
10                    20.197148
11                    16.170638
12                    15.887671
13                    16.583536
14                    16.773669
15                    16.165037
16                    17.665606
17                    15.694469
18                    19.310397
19                    17.177343
20                    16.473447
21                    22.902379

## Q3. The results for the entire dataframe contains the average for all 21 departments, where as the subset only contains the average for 8 departments. Also, the highest average in the subset is department_id 16 (19.46) but in the full dataframe, we can see that department_id 21 has the highest average of 22.90. Similarly, the lowest average in the subset is department_id 17 with 11.29 but in the full dataframe, department_id 5 has the lowest average (15.22).

In [6]:
# Q4.Create user id grouping, performed aggregate function and add to new column 'max order' to ords_prods_merge dataframe

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['user_order_number'].transform(np.max)

C:\Users\Frederick\AppData\Local\Temp\ipykernel_9272\1891934290.py:3: FutureWarning: The provided callable <function amax at 0x0000029214DB7C40> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['user_order_number'].transform(np.max)


In [7]:
# Q4.Create a loyalty flag for loyal customer

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

C:\Users\Frederick\AppData\Local\Temp\ipykernel_9272\151755974.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Loyal customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


In [8]:
# Q4.Create a loyalty flag for Regular customer

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [9]:
# Q4.Create a loyalty flag for New customer

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [10]:
# Checking output on the new loyalty flags with user_id, loyalty_flag and user_order_number columns

ords_prods_merge[['user_id', 'loyalty_flag', 'user_order_number']].head(60)

user_id      loyalty_flag  user_order_number
0         1      New customer                  1
1         1      New customer                  2
2         1      New customer                  3
3         1      New customer                  4
4         1      New customer                  5
5         1      New customer                  6
6         1      New customer                  7
7         1      New customer                  8
8         1      New customer                  9
9         1      New customer                 10
10       15  Regular customer                 15
11       15  Regular customer                 17
12       15  Regular customer                 18
13       15  Regular customer                 21
14       15  Regular customer                 22
15       19      New customer                  2
16       19      New customer                  5
17       19      New customer                  7
18       21  Regular customer                 10
19       31  Regular customer                 10
20       31  Regular customer                 17
21       43  Regular customer                  6
22       43  Regular customer                  9
23       52  Regular customer                  1
24       52  Regular customer                  2
25       52  Regular customer                  3
26       52  Regular customer                  5
27       52  Regular customer                  7
28       52  Regular customer                  9
29       52  Regular customer                 10
30       52  Regular customer                 11
31       52  Regular customer                 12
32       52  Regular customer                 15
33       52  Regular customer                 17
34       52  Regular customer                 18
35       52  Regular customer                 19
36       52  Regular customer                 21
37       67  Regular customer                  1
38       67  Regular customer                  2
39       67  Regular customer                  3
40       67  Regular customer                  4
41       67  Regular customer                  5
42       67  Regular customer                  6
43       67  Regular customer                  8
44       67  Regular customer                  9
45       67  Regular customer                 10
46       67  Regular customer                 11
47       67  Regular customer                 12
48       67  Regular customer                 13
49       67  Regular customer                 14
50       67  Regular customer                 16
51       67  Regular customer                 18
52       67  Regular customer                 19
53       67  Regular customer                 21
54       67  Regular customer                 22
55       67  Regular customer                 23
56       81      New customer                  3
57       81      New customer                  5
58       82  Regular customer                  1
59       82  Regular customer                  3

In [11]:
# Q5. Find aggregated mean of 'prices' column grouped by loyalty_flag 

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

## Q5. From the above table, we can see that the spending habits are different for loyal customers to that of regular and new customers. Loyal customers tended to buy products with lower prices (average of 10.37) where as new customers tended to buy the more expensive products (average 13.29), followed by regular customers (average 12.50). 

In [12]:
# Q6.Create user id grouping, performed aggregate function on prices and add to new column 'avg_price' to ords_prods_merge dataframe

ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Frederick\AppData\Local\Temp\ipykernel_9272\2784604785.py:3: FutureWarning: The provided callable <function mean at 0x0000029214DBC9A0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [13]:
# Q6. Checking output on the new column 'avg_price'

ords_prods_merge.head(15)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  user_order_number  \
0              0             0   2539329        1                  1   
1              1             1   2398795        1                  2   
2              2             2    473747        1                  3   
3              3             3   2254736        1                  4   
4              4             4    431534        1                  5   
5              5             5   3367565        1                  6   
6              6             6    550135        1                  7   
7              7             7   3108588        1                  8   
8              8             8   2295261        1                  9   
9              9             9   2550362        1                 10   
10           144           144   2968173       15                 15   
11           146           146   1870022       15                 17   
12           147           147   1911383       15                 18   
13           150           150   2715276       15                 21   
14           151           151    487368       15                 22   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
5                    2                  7                    19.0         196   
6                    1                  9                    20.0         196   
7                    1                 14                    14.0         196   
8                    1                 16                     0.0         196   
9                    4                  8                    30.0         196   
10                   1                  9                     7.0         196   
11                   2                 16                     8.0         196   
12                   2                 11                     7.0         196   
13                   1                  9                     7.0         196   
14                   1                 10                    14.0         196   

    add_to_cart_order  ...  department_id prices  exists     busiest_day  \
0                   1  ...              7    9.0    both  Regularly busy   
1                   1  ...              7    9.0    both  Regularly busy   
2                   1  ...              7    9.0    both  Regularly busy   
3                   1  ...              7    9.0    both      Least busy   
4                   1  ...              7    9.0    both      Least busy   
5                   1  ...              7    9.0    both  Regularly busy   
6                   1  ...              7    9.0    both  Regularly busy   
7                   2  ...              7    9.0    both  Regularly busy   
8                   4  ...              7    9.0    both  Regularly busy   
9                   1  ...              7    9.0    both      Least busy   
10                  2  ...              7    9.0    both  Regularly busy   
11                  6  ...              7    9.0    both  Regularly busy   
12                  1  ...              7    9.0    both  Regularly busy   
13                  1  ...              7    9.0    both  Regularly busy   
14                  1  ...              7    9.0    both  Regularly busy   

      price_range_loc  busiest_days  busiest_period_of_day max_order  \
0   Mid-range product  Regular days         Average orders        10   
1   Mid-range product  Slowest days         Average orders        10   
2   Mid-range product  Slowest days            Most orders        10   
3   Mid-range product  Slowest days         Avera

In [14]:
# Q6.Create a spending flag for low spender

ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'

C:\Users\Frederick\AppData\Local\Temp\ipykernel_9272\3557069658.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Low spender' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spending_flag'] = 'Low spender'


In [15]:
# Q6.Create a spending flag for High spender

ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [16]:
# Checking output on the new spending flags with user_id, spending_flag and avg_price columns

ords_prods_merge[['user_id', 'spending_flag', 'avg_price']].head(100)

user_id spending_flag  avg_price
0         1   Low spender   6.367797
1         1   Low spender   6.367797
2         1   Low spender   6.367797
3         1   Low spender   6.367797
4         1   Low spender   6.367797
..      ...           ...        ...
95      360  High spender  10.006250
96      377   Low spender   8.496552
97      387   Low spender   7.396610
98      420   Low spender   7.387805
99      420   Low spender   7.387805

[100 rows x 3 columns]

In [17]:
# Q7.Create user id grouping, performed aggregate function on days_since_prior_order and add to new column 'median_days' to ords_prods_merge dataframe

ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Frederick\AppData\Local\Temp\ipykernel_9272\3397459266.py:3: FutureWarning: The provided callable <function median at 0x0000029214EFA5C0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [18]:
# Q7. Checking output on the new column 'median_days_since_prior_order'

ords_prods_merge.head(15)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  user_order_number  \
0              0             0   2539329        1                  1   
1              1             1   2398795        1                  2   
2              2             2    473747        1                  3   
3              3             3   2254736        1                  4   
4              4             4    431534        1                  5   
5              5             5   3367565        1                  6   
6              6             6    550135        1                  7   
7              7             7   3108588        1                  8   
8              8             8   2295261        1                  9   
9              9             9   2550362        1                 10   
10           144           144   2968173       15                 15   
11           146           146   1870022       15                 17   
12           147           147   1911383       15                 18   
13           150           150   2715276       15                 21   
14           151           151    487368       15                 22   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
5                    2                  7                    19.0         196   
6                    1                  9                    20.0         196   
7                    1                 14                    14.0         196   
8                    1                 16                     0.0         196   
9                    4                  8                    30.0         196   
10                   1                  9                     7.0         196   
11                   2                 16                     8.0         196   
12                   2                 11                     7.0         196   
13                   1                  9                     7.0         196   
14                   1                 10                    14.0         196   

    add_to_cart_order  ...  exists     busiest_day    price_range_loc  \
0                   1  ...    both  Regularly busy  Mid-range product   
1                   1  ...    both  Regularly busy  Mid-range product   
2                   1  ...    both  Regularly busy  Mid-range product   
3                   1  ...    both      Least busy  Mid-range product   
4                   1  ...    both      Least busy  Mid-range product   
5                   1  ...    both  Regularly busy  Mid-range product   
6                   1  ...    both  Regularly busy  Mid-range product   
7                   2  ...    both  Regularly busy  Mid-range product   
8                   4  ...    both  Regularly busy  Mid-range product   
9                   1  ...    both      Least busy  Mid-range product   
10                  2  ...    both  Regularly busy  Mid-range product   
11                  6  ...    both  Regularly busy  Mid-range product   
12                  1  ...    both  Regularly busy  Mid-range product   
13                  1  ...    both  Regularly busy  Mid-range product   
14                  1  ...    both  Regularly busy  Mid-range product   

    busiest_days  busiest_period_of_day  max_order      loyalty_flag  \
0   Regular days         Average orders         10      New customer   
1   Slowest days         Average orders         10      New customer   
2   Slowest days            Most orders         10      New customer   
3   Slowest days         Average orders         10      New customer   
4   Slowest days         

In [19]:
# Q7.Create a frequency flag for Non-frequent customer

ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

C:\Users\Frederick\AppData\Local\Temp\ipykernel_9272\1887152545.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Non-frequent customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'


In [20]:
# Q7.Create a frequency flag for Regular customer

ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] <= 20) & (ords_prods_merge['median_days_since_prior_order'] > 10), 'frequency_flag'] = 'Regular customer'

In [21]:
# Q7.Create a frequency flag for Frequent customer

ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [22]:
# Q7.Checking output on the new frequency flags with user_id, frequency_flag and median_days_since_prior_order columns

ords_prods_merge[['user_id', 'frequency_flag', 'median_days_since_prior_order']].head(60)

user_id         frequency_flag  median_days_since_prior_order
0         1  Non-frequent customer                           20.5
1         1  Non-frequent customer                           20.5
2         1  Non-frequent customer                           20.5
3         1  Non-frequent customer                           20.5
4         1  Non-frequent customer                           20.5
5         1  Non-frequent customer                           20.5
6         1  Non-frequent customer                           20.5
7         1  Non-frequent customer                           20.5
8         1  Non-frequent customer                           20.5
9         1  Non-frequent customer                           20.5
10       15      Frequent customer                           10.0
11       15      Frequent customer                           10.0
12       15      Frequent customer                           10.0
13       15      Frequent customer                           10.0
14       15      Frequent customer                           10.0
15       19      Frequent customer                            6.0
16       19      Frequent customer                            6.0
17       19      Frequent customer                            6.0
18       21      Frequent customer                            7.5
19       31      Frequent customer                            6.0
20       31      Frequent customer                            6.0
21       43      Frequent customer                           10.0
22       43      Frequent customer                           10.0
23       52      Frequent customer                            8.0
24       52      Frequent customer                            8.0
25       52      Frequent customer                            8.0
26       52      Frequent customer                            8.0
27       52      Frequent customer                            8.0
28       52      Frequent customer                            8.0
29       52      Frequent customer                            8.0
30       52      Frequent customer                            8.0
31       52      Frequent customer                            8.0
32       52      Frequent customer                            8.0
33       52      Frequent customer                            8.0
34       52      Frequent customer                            8.0
35       52      Frequent customer                            8.0
36       52      Frequent customer                            8.0
37       67      Frequent customer                            7.0
38       67      Frequent customer                            7.0
39       67      Frequent customer                            7.0
40       67      Frequent customer                            7.0
41       67      Frequent customer                            7.0
42       67      Frequent customer                            7.0
43       67      Frequent customer                            7.0
44       67      Frequent customer                            7.0
45       67      Frequent customer                            7.0
46       67      Frequent customer                            7.0
47       67      Frequent customer                            7.0
48       67      Frequent customer                            7.0
49       67      Frequent customer                            7.0
50       67      Frequent customer                            7.0
51       67      Frequent customer                            7.0
52       67      Frequent customer                            7.0
53       67      Frequent customer                            7.0
54       67      Frequent customer                            7.0
55       67      Frequent customer                            7.0
56       81  Non-frequent customer                           30.0
57       81  Non-frequent customer                           30.0
58       82       Regular customer                           11.5
59       82       Regular customer          

In [23]:
# Assign missing values to suspected outliers in prices

ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [24]:
# Check max value on prices

ords_prods_merge['prices'].max()

25.0

In [25]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_flags.pkl'))

In [27]:
# Check output for loyalty flag

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [28]:
# Check output for spending flag

ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [29]:
# Check output for frequency flag

ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64